prov/util/src/util_recv.c

```diff
@@ -0,0 +1,203 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_util.h>
+
+static int ofi_match_recv_any(struct dlist_entry *item, const void *arg)
+{
+	OFI_UNUSED(item);
+	OFI_UNUSED(arg);
+	return 1;
+}
+
+static int ofi_match_recv_addr(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_match_attr *match_attr = (struct ofi_match_attr *)arg;
+	struct ofi_recv_entry *recv_entry =
+		container_of(item, struct ofi_recv_entry, entry);
+
+	return ofi_match_addr(recv_entry->addr, match_attr->addr);
+}
+
+static int ofi_match_recv_tag(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_match_attr *match_attr = (struct ofi_match_attr *)arg;
+	struct ofi_recv_entry *recv_entry =
+		container_of(item, struct ofi_recv_entry, entry);
+
+	return ofi_match_tag(recv_entry->tag, recv_entry->ignore, match_attr->tag);
+}
+
+static int ofi_match_recv_addr_tag(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_match_attr *match_attr = (struct ofi_match_attr *)arg;
+	struct ofi_recv_entry *recv_entry =
+		container_of(item, struct ofi_recv_entry, entry);
+
+	return ofi_match_addr(recv_entry->addr, match_attr->addr) &&
+		ofi_match_tag(recv_entry->tag, recv_entry->ignore, match_attr->tag);
+}
+
+static int ofi_match_rx_any(struct dlist_entry *item, const void *arg)
+{
+	OFI_UNUSED(item);
+	OFI_UNUSED(arg);
+	return 1;
+}
+
+static int ofi_match_rx_addr(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_recv_entry *recv_entry = (struct ofi_recv_entry *)arg;
+	struct ofi_rx_entry *rx_entry =
+		container_of(item, struct ofi_rx_entry, entry);
+
+	return ofi_match_addr(rx_entry->addr, recv_entry->addr);
+}
+
+static int ofi_match_rx_tag(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_recv_entry *recv_entry = (struct ofi_recv_entry *)arg;
+	struct ofi_rx_entry *rx_entry =
+		container_of(item, struct ofi_rx_entry, entry);
+
+	return ofi_match_tag(recv_entry->tag, recv_entry->ignore, rx_entry->tag);
+}
+
+static int ofi_match_rx_addr_tag(struct dlist_entry *item, const void *arg)
+{
+	struct ofi_recv_entry *recv_entry = (struct ofi_recv_entry *)arg;
+	struct ofi_rx_entry *rx_entry =
+		container_of(item, struct ofi_rx_entry, entry);
+
+	return ofi_match_addr(recv_entry->addr, rx_entry->addr) &&
+		ofi_match_tag(recv_entry->tag, recv_entry->ignore, rx_entry->tag);
+}
+
+int ofi_reprocess_directed_recvs(struct ofi_recv_queue *recv_queue)
+{
+	// TODO
+	return -FI_ENOSYS;
+}
+
+struct ofi_recv_entry *
+ofi_match_recv(struct ofi_recv_queue *recv_queue,
+	       struct ofi_match_attr *match_attr,
+	       char *data, size_t len, void *context)
```

Comment:
I can update the function name. The objective of the function is to find a matching recv and if it cannot find one, en-queue the received message / completion entry to the unexpected queue. Why do we need to separate the latter part? I'm assuming any software tag matching solution would want to maintain an unexpected queue.

Reply:
Queuing to the unexpected queue is a policy decision that the provider should own.  In the case of something like RxD, it may want to drop unexpected messages to reduce memory pressure.  In the worst case, the caller ends up with an additional 2 lines of code. if (no match) queue.

Reply:
Got it. I'll create separate functions.